migrations/20210204190556-add-slug-to-projects.js

```diff
@@ -0,0 +1,100 @@
+const async = require('async');
+
+function slugify(title) {
+  return title
+    .replace(/[^a-zA-Z\d\s_()-./\\]/g, '')
+    .replace(/(\s|_|\(|\)|\/|\\|\.)+/g, '-')
+    .toLowerCase();
+}
+
+async function createSlug(id, title, collection) {
+  const slug = slugify(title);
+  let realSlug;
+  let count = 0;
+  let postfix = 0;
+  await async.doWhilst(
+    cb => {
+      realSlug = postfix === 0 ? slug : `${slug}-${postfix + 1}`;
+      collection
+        .countDocuments({
+          slug: realSlug,
+          _id: { $ne: id },
+        })
+        .then(_count => {
+          count = _count;
+          cb();
+        })
+        .catch(err => {
+          cb(err);
+        });
+      postfix += 1;
+    },
+    testCb => {
+      testCb(null, count > 0);
+    },
+  );
+  return realSlug;
+}
+
+module.exports = {
+  async up(db, client) {
+    // add slug to dacs
+    const dacs = await db
+      .collection('dacs')
+      .find({})
+      .project({ title: true })
+      .toArray();
+    await async.eachLimit(dacs, 1, (dac, cb) => {
+      const { title } = dac;
+      createSlug(dac._id, title, db.collection('dacs')).then(slug => {
+        db.collection('dacs')
+          .updateOne({ _id: dac._id }, { $set: { slug } })
+          .then(() => {
+            cb();
+          });
+      });
+    });
+    // add slug to campaigns
+    const campaigns = await db
+      .collection('campaigns')
+      .find({})
+      .project({ title: true })
+      .toArray();
+    await async.eachLimit(campaigns, 1, (campaign, cb) => {
+      const { title } = campaign;
+      createSlug(campaign._id, title, db.collection('campaigns')).then(slug => {
+        db.collection('campaigns')
+          .updateOne({ _id: campaign._id }, { $set: { slug } })
+          .then(() => {
+            cb();
+          });
+      });
+    });
+
+    // add slug to milestones
+    const milestones = await db
+      .collection('milestones')
+      .find({})
+      .project({ title: true })
+      .toArray();
+    await async.eachLimit(milestones, 1, (milestone, cb) => {
+      const { title } = milestone;
+      createSlug(milestone._id, title, db.collection('milestones')).then(slug => {
+        db.collection('milestones')
+          .updateOne({ _id: milestone._id }, { $set: { slug } })
+          .then(() => {
+            cb();
+          });
+      });
+    });
+    await db.collection('dacs').createIndex({ slug: 1 }, { unique: true });
```

Comment:
@mdehghani 
I ran migration on beta dump and at the end of script there many dacs, campaigns and milestones that doesnt have  `slug` so when adding unique index  to them there are multiple documents that have `null` value in index and getting error, so migration script doesnt work well :

```
ERROR: Could not migrate up 20210204190556-add-slug-to-projects.js: Index build failed: 8089dd78-fce3-4ded-ac34-d17621175985: Collection giveth-beta.dacs ( d59b10a4-209d-4ae1-97ef-a376e276c547 ) :: caused by :: E11000 duplicate key error collection: giveth-beta.dacs index: slug_1 dup key: { slug: null }
```

Reply:
Fixed